commit df38157d94662a64e2f83aa8a0110fd1ee7c4776
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Thu Apr 27 11:55:06 2023 -0400

    In hstore_plpython, avoid crashing when return value isn't a mapping.
    
    Python 3 changed the behavior of PyMapping_Check(), breaking the
    test in plpython_to_hstore() that verifies whether a function result
    to be transformed is acceptable.  A backwards-compatible fix is to
    first verify that the object doesn't pass PySequence_Check().
    
    Perhaps accidentally, our other uses of PyMapping_Check() already
    follow uses of PySequence_Check(), so that no other bugs were
    created by this change.
    
    Per bug #17908 from Alexander Lakhin.  Back-patch to all supported
    branches.
    
    Dmitry Dolgov and Tom Lane
    
    Discussion: https://postgr.es/m/17908-3f19a125d56a11d6@postgresql.org

diff --git a/contrib/hstore_plpython/expected/hstore_plpython.out b/contrib/hstore_plpython/expected/hstore_plpython.out
index bf238701fec..5fb56a2f65d 100644
--- a/contrib/hstore_plpython/expected/hstore_plpython.out
+++ b/contrib/hstore_plpython/expected/hstore_plpython.out
@@ -30,10 +30,21 @@ INFO:  [('aa', 'bb'), ('cc', None)]
  test1n 
 --------
       2
 (1 row)
 
+-- test that a non-mapping result is correctly rejected
+CREATE FUNCTION test1bad() RETURNS hstore
+LANGUAGE plpython3u
+TRANSFORM FOR TYPE hstore
+AS $$
+return "foo"
+$$;
+SELECT test1bad();
+ERROR:  not a Python mapping
+CONTEXT:  while creating return value
+PL/Python function "test1bad"
 -- test hstore[] -> python
 CREATE FUNCTION test1arr(val hstore[]) RETURNS int
 LANGUAGE plpython3u
 TRANSFORM FOR TYPE hstore
 AS $$
diff --git a/contrib/hstore_plpython/hstore_plpython.c b/contrib/hstore_plpython/hstore_plpython.c
index 961579a5ea0..310f63c30d4 100644
--- a/contrib/hstore_plpython/hstore_plpython.c
+++ b/contrib/hstore_plpython/hstore_plpython.c
@@ -125,11 +125,17 @@ plpython_to_hstore(PG_FUNCTION_ARGS)
 	PyObject   *volatile items;
 	Py_ssize_t	pcount;
 	HStore	   *volatile out;
 
 	dict = (PyObject *) PG_GETARG_POINTER(0);
-	if (!PyMapping_Check(dict))
+
+	/*
+	 * As of Python 3, PyMapping_Check() is unreliable unless one first checks
+	 * that the object isn't a sequence.  (Cleaner solutions exist, but not
+	 * before Python 3.10, which we're not prepared to require yet.)
+	 */
+	if (PySequence_Check(dict) || !PyMapping_Check(dict))
 		ereport(ERROR,
 				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
 				 errmsg("not a Python mapping")));
 
 	pcount = PyMapping_Size(dict);
diff --git a/contrib/hstore_plpython/sql/hstore_plpython.sql b/contrib/hstore_plpython/sql/hstore_plpython.sql
index a9cfbbe13e2..ebd61e6c467 100644
--- a/contrib/hstore_plpython/sql/hstore_plpython.sql
+++ b/contrib/hstore_plpython/sql/hstore_plpython.sql
@@ -25,10 +25,21 @@ return len(val)
 $$;
 
 SELECT test1n('aa=>bb, cc=>NULL'::hstore);
 
 
+-- test that a non-mapping result is correctly rejected
+CREATE FUNCTION test1bad() RETURNS hstore
+LANGUAGE plpython3u
+TRANSFORM FOR TYPE hstore
+AS $$
+return "foo"
+$$;
+
+SELECT test1bad();
+
+
 -- test hstore[] -> python
 CREATE FUNCTION test1arr(val hstore[]) RETURNS int
 LANGUAGE plpython3u
 TRANSFORM FOR TYPE hstore
 AS $$
